src/ngraph/op/experimental/batch_dot.cpp

```diff
@@ -0,0 +1,139 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#include "batch_dot.hpp"
+#include "ngraph/dimension.hpp"
+#include "ngraph/log.hpp"
+#include "ngraph/op/experimental/dyn_reshape.hpp"
+#include "ngraph/op/reshape.hpp"
+#include "ngraph/util.hpp"
+
+using namespace std;
+using namespace ngraph;
+
+op::BatchDot::BatchDot(const shared_ptr<Node>& arg0,
+                       const shared_ptr<Node>& arg1,
+                       bool transpose_arg0,
+                       bool transpose_arg1)
+    : Op("BatchDot", check_single_output_args({arg0, arg1}))
+    , m_transpose_arg0(transpose_arg0)
+    , m_transpose_arg1(transpose_arg1)
+{
+    constructor_validate_and_infer_types();
+}
+
+shared_ptr<Node> op::BatchDot::copy_with_new_args(const NodeVector& new_args) const
+{
+    if (new_args.size() != 2)
+    {
+        throw ngraph_error("Incorrect number of new arguments");
+    }
+
+    return make_shared<BatchDot>(
+        new_args.at(0), new_args.at(1), m_transpose_arg0, m_transpose_arg1);
+}
+
+void op::BatchDot::validate_and_infer_types()
+{
+    // Check input types
+    auto arg0_et = get_input_element_type(0);
+    auto arg1_et = get_input_element_type(1);
+    NODE_VALIDATION_CHECK(this,
+                          arg0_et.compatible(arg1_et),
+                          "Inputs arg0 and arg1 must have compatible element type.");
+    // Check input shapes
+    const PartialShape& arg0_shape = get_input_partial_shape(0);
+    const PartialShape& arg1_shape = get_input_partial_shape(1);
+    NODE_VALIDATION_CHECK(this,
+                          arg0_shape.rank().compatible(3),
+                          "Input arg0 shape must have rank 3, got ",
+                          arg0_shape.rank(),
+                          ".");
+    NODE_VALIDATION_CHECK(this,
+                          arg1_shape.rank().compatible(3),
+                          "Input arg1 shape must have rank 3, got ",
+                          arg1_shape.rank(),
+                          ".");
+
+    size_t dot_dim_arg0 = (m_transpose_arg0) ? 1 : 2;
+    size_t dot_dim_arg1 = (m_transpose_arg1) ? 2 : 1;
+
+    // We expect output shape always have rank 3
+    PartialShape output_shape(PartialShape::dynamic(3));
+
+    // Construct output shape with more information if avalible.
+    if (arg0_shape.rank().is_static() && arg1_shape.rank().is_static() &&
```

Comment:
You can rewrite this condition as:

`if (arg0_shape.rank().same_scheme(3) && arg1_shape.rank().same_scheme(3))`.

(The nomenclature here is odd, perhaps we should change it. Perhaps something like `necessarily_eq`?)

Reply:
Fixed. It looks to the logic in `same_scheme()` is essentially what I would call `equal`.

Reply:
What I was thinking is that it's important to distinguish between the case where s1 and s2 are necessarily the same shape at runtime (meaning, really, that both are static and equal), and the case where s1 and s2 represent quantities that statically aren't yet known to be different but could be at runtime (e.g. a [?,2] over here and [?,2] over there could respectively wind up being [3,2] and [4,2] in a given iteration).